packages/embarkjs/.yarnrc

```diff
@@ -0,0 +1,2 @@
+--*.scripts-prepend-node-path true
+--install.check-files true
```

Comment:
Same here, that can probably be at the root?

Reply:
As a general principle, to know what should be in a package vs. extracted to root:

*If you copy `packages/[pkg]` outside of the monorepo and do `yarn|npm install` and run other scripts like `build`, etc., what needs to be there for it work as expected?*

On that basis, that's why devDependencies have to be fully specified in each of `/packages/*`, and things like `.npmrc`, babel configuration, etc. need to be in place depending on what's involved in the package (`@types/embark` package doesn't use babel, for example, and doesn't even need a `build` script).

When answering the question, we can make a mental exception for things like private packages in `packages/*` that would be unavailable from the registry, and we don't need to consider things like `.gitignore` that are orthogonal to node/npm/yarn.

The monorepo is doing essentially two things for us:

1) The yarn workspace provides time/space savings re: `node_modules` among `packages/*`. We can increase the savings if `packages/*` use the same versions of babel, etc. The workspace also sets up a bunch of symlinks and/or shim scripts so that `npm run` / `yarn run` work as expected when it comes to their automatically putting `./node_modules/.bin` on the PATH.

2) Lerna can invoke all the scripts in `packages/*` based on common names like `build`, `lint`, `test`, etc. `lerna run` is basically just `cd`'ing into a package directory and running the named script.  There is also `lerna exec` for running arbitrary commands across all/scoped packages.

---

Can we cut down on boilerplate? For example, could we have a common babel config and the babel packages as root devDeps only?

We can do that but it will require rethinking how the root `package.json` scripts work. Right now, the root doesn't anything about how a package chooses to implement `build` or other scripts.  In the alternative approach, it's root that knows how to build the various packages and the packages don't know how to build themselves.

At this point, I am undecided about which approach is better, they both have trade-offs. My gut instinct is that putting build-logic, test-logic, etc. in the individual packages provides more flexibility (better encapsulation) at the cost of more boilerplate in each package.

I'm very interested in hearing alternative views on this; there may be aspects of practical use of lerna that I have misunderstood.

Reply:
> If you copy packages/[pkg] outside of the monorepo and do yarn|npm install and run other scripts like build, etc., what needs to be there for it work as expected?

It is fine for me if nothing works. I would expect all babel configuration|or any build configuration to be out of the scope of the package.
IMO what should be in packages are things that make them different from each other.

For example this repo: https://github.com/0xProject/0x-monorepo
You can see the `tsconfig` is at root
That helps to follow a pattern that I like: Convention over Configuration
All embark packages are built the same way.

For the devDeps, I keep them inside each packages so they appear nicely on tool like https://www.npmjs.com/

To summarize my opinion:
- Keep deps/devDeps per package
- Extract configuration out of each package if possible

Reply:
I want to clarify re: 

> If you copy packages/[pkg] outside of the monorepo and do yarn|npm install and run other scripts like build, etc., what needs to be there for it work as expected?

I don't mean it as an ideal I have. I meant that's the actual operating principle for how the `package.json` scripts work that are in place currently.

Reply:
I will look at 0x-monorepo to understand how that arrangement works.

The docs for tsconfig  say that tsc will look up the directory tree recursively for tsconfig, so in that case we can definitely move it from `packages/embark` to the root without needing to change anything else, and we don't need it in other places either, e.g. `packages/embark-typings`.

I will double-check re: `.npmrc`, `babel.config.js`, etc. to see which ones can be auto-located higher in the directory structure.

Reply:
> That helps to follow a pattern that I like: Convention over Configuration.

I like that too, it's what I've aimed for with the scripts setup in the monorepo — just provide a `build`, `test`, `ci`, etc. scripts in `packages/whatever` and the monorepo will do the right thing in the right order.

> All embark packages are built the same way.

Definitely where possible!  `embark-ui` has a unique setup because of CRA, but hopefully most of our packages will have a common setup re: babel, etc.

Reply:
Looks like there is a way to have a common `babel.config.js` at the root of the monorepo while still having each package spec its `build` script, i.e. we don't need to refactor the `build` script logic at the root level.

https://babeljs.io/docs/en/config-files#cli

`embarkjs` needs its own `babel.config.js` (just renamed from `.babelrc`, but for many of `packages/embark-*` a common config will work great.

Reply:
It's possible to cleanup `.yarnrc` a bit because it auto-merges up the directory tree (verified by experiment), will submit a PR shortly. Unfortunately, `.npmrc` doesn't work like that (verified), so each package needs one.

Reply:
`.yarnrc` has been cleaned up in #1298.

Reply:
A root babel config is introduced in #1299.  root tslint and tsconfig will be introduced in #1290 — I'm about to push some commits into that one.